src/test/scala/fm/xml/TestMultiXmlReader.scala

```diff
@@ -0,0 +1,438 @@
+  <!-- hello -->
+  <foo><bar>asdasd</bar><part><uniqueId>SKIPME</uniqueId></part></foo>
+  <items>
+    <part>
+      <!-- hello -->
+      <uniqueId>ONE</uniqueId>
+      <source>foo</source>
+      <name>Simple Part One</name>
+    </part>
+    <price>
+      <!-- hello -->
+      <uniqueId>ONE</uniqueId>
+      <price>9.99</price>
+    </price>
+    <foo><part><uniqueId>SKIPME</uniqueId></part></foo>
+    <part>
+      <uniqueId>TWO</uniqueId>
+      <source>foo</source>
+      <name>Simple Part Two</name>
+      <!-- hello -->
+    </part>
+    <price>
+      <uniqueId>TWO</uniqueId>
+      <price>19.99</price>
+      <!-- hello -->
+    </price>
+  </items>
+  <!-- hello -->
+</feed>
+"""
+  test("Multi-Nested Part - Reading") {
+    val reader = makeReader(multiNestedPartXml, paths=Seq(NestedItemsPartPath, NestedItemsPricePath))
+    reader.hasNext should equal(true)
+    checkOnePart(reader.next)
+
+    reader.hasNext should equal(true)
+    checkOnePrice(reader.next)
+
+    reader.hasNext should equal(true)
+    checkTwoPart(reader.next)
+
+    reader.hasNext should equal(true)
+    checkTwoPrice(reader.next)
+
+    reader.hasNext should equal(true)
+    checkOnePart(reader.next)
+
+    reader.hasNext should equal(true)
+    checkOnePrice(reader.next)
+
+    reader.hasNext should equal(true)
+    checkTwoPart(reader.next)
+
+    reader.hasNext should equal(true)
+    checkTwoPrice(reader.next)
+
+    reader.hasNext should equal(false)
+  }
+
+  private val emptyFeedXml = """<?xml version='1.0' encoding='UTF-8'?>
+<feed>
+</feed>
+"""
+
+  test("Empty Feed") {
+    val reader = makeReader(emptyFeedXml, paths=Seq(NestedItemsPartPath, NestedItemsPricePath))
+    reader.hasNext should equal(false)
+  }
+
+  test("Feed that has elements but not what we are looking for") {
+    val reader = makeReader(nestedPartXml, paths=Seq(NestedItemsPartElementNotExistPath, NestedItemsPriceElementNotExistPath))
+    reader.hasNext should equal(false)
+  }
+
+  trait MyPathValue extends AnyRef
+
+  case class PartPathValue(value: SimpleFeedPart) extends MyPathValue
+  case class PricePathValue(value: SimpleFeedPrice) extends MyPathValue
+
+  case class NestedItemsPartPathValue(value: SimpleFeedPart) extends MyPathValue
+  case class NestedItemsPricePathValue(value: SimpleFeedPrice) extends MyPathValue
+
+  case class NestedItemsPartElementNotExistPathValue(value: SimpleFeedPart) extends MyPathValue
+  case class NestedItemsPriceElementNotExistPathValue(value: SimpleFeedPrice) extends MyPathValue
+
+  case class DeepNestedPartPathValue(value: SimpleFeedPart) extends MyPathValue
+  case class DeepNestedPricePathValue(value: SimpleFeedPrice) extends MyPathValue
+
+  case class OnePartPathValue(value: SimpleFeedPart) extends MyPathValue
+  case class OnePricePathValue(value: SimpleFeedPrice) extends MyPathValue
+
+  case class TwoPartPathValue(value: SimpleFeedPart) extends MyPathValue
+  case class TwoPricePathValue(value: SimpleFeedPrice) extends MyPathValue
```

Comment:
This whole thing seems like an odd pattern...

Reply:
I agree, but the only way you can get an exhaustive pattern match error is if the base value that is returned is sealed.  We also want the values to be specific the the item-path, and not just based upon the return type.